src/dtree.ml

```diff
@@ -0,0 +1,499 @@
+      (List.map (fun { lhs ; _ } -> lhs) clauses) ;
+    F.fprintf oc "@.}@,"
+
+  (** [of_rules r] creates the initial pattern matrix from a list of rewriting
+      rules. *)
+  let of_rules : Terms.rule list -> t = fun rs ->
+    let r2r r =
+      let term_pos = List.to_seq r.Terms.lhs |> Subterm.tag |> Array.of_seq in
+      { lhs = term_pos ; rhs = r.Terms.rhs ; env_builder = [] } in
+    { clauses = List.map r2r rs ; var_met = 0 }
+
+  (** [is_empty m] returns whether matrix [m] is empty. *)
+  let is_empty : t -> bool = fun m -> m.clauses = []
+
+  (** [get_col n m] retrieves column [n] of matrix [m]. *)
+  let get_col : int -> t -> component list = fun ind { clauses ; _ } ->
+    List.map (fun { lhs ; _ } -> lhs.(ind)) clauses
+
+  (** [score c] returns the score heuristic for column [c]. *)
+  let rec score : component list -> int = function
+    | []                              -> 0
+    | (x, _) :: xs when is_treecons x -> score xs
+    | _ :: xs                         -> succ (score xs)
+
+  (** [pick_best_among m c] returns the index of the best column of matrix [m]
+      among columns [c] according to a heuristic. *)
+  let pick_best_among : t -> int array -> int = fun mat columns->
+    let wild_pc = Array.map (fun ci ->
+      let c = get_col ci mat in score c) columns in
+    Array.argmax (<=) wild_pc
+
+  (** [yield m] yields a rule to be applied. *)
+  let yield : t -> rule option = fun { clauses ; _ } ->
+    let is_exhausted { lhs ; _ } =
+      Array.for_all (fun (e, _) -> not (is_treecons e)) lhs in
+    List.find_opt is_exhausted clauses
+
+  (** [can_switch_on p k] returns whether a switch can be carried out on
+      column [k] in matrix [p] *)
+  let can_switch_on : t -> int -> bool = fun { clauses ; _ } k ->
+    let rec loop : rule list -> bool = function
+      | []      -> false
+      | r :: rs -> if is_treecons (fst r.lhs.(k)) then true else loop rs in
+    loop clauses
+
+  (** [discard_cons_free m] returns the list of indexes of columns containing
+      terms that can be matched against (discard constructor-free columns). *)
+  let discard_cons_free : t -> int array = fun m ->
+    let ncols = List.extremum (>)
+      (List.map (fun { lhs ; _ } -> Array.length lhs) m.clauses) in
+    let switchable = List.init ncols (can_switch_on m) in
+    let switchable2ind i e = if e then Some(i) else None in
+    switchable |> List.filteri_map switchable2ind |> Array.of_list
+
+  (** [get_cons l] extracts, sorts and uniqify terms that are tree
+      constructors in [l].  The actual tree constructor (of type
+      {!type:treecons}) is returned along the original term. *)
+  let get_cons : term list -> (treecons * term) list = fun telst ->
+    let keep_treecons e =
+      if is_treecons e then Some(treecons_of_term e, e) else None in
+    let tc_fst_cmp (tca, _) (tcb, _) = tc_compare tca tcb in
+    telst |> List.filter_map keep_treecons |> List.sort_uniq tc_fst_cmp
```

Comment:
Why do you need to call List.sort_uniq?

Reply:
What if you have a list with (tc,t1) and (tc,t2) with t1<>t2?

Reply:
We may have several lines having the same constructor, we don't want to specialise several times against the same constructor

Reply:
The case you mentioned shouldn't happen, or it means that `tc` is ill defined

Reply:
Right. And what about List.sort_uniq?

Reply:
If we have the matrix
```
[ c &x &y ]
[ c a  &y ]
[ d &x _  ]
```
we want to have a specialisation regarding `c`, and another one regarding `d`.  If we don't `sort_uniq`, we'll have two specialisations regarding `c`.

Reply:
I see.